### DISCUSSION TITLE
MQTT PubSubClient.h compatibility

### DISCUSSION THREAD

**Discussion by kyleti** [ssilverman/QNEthernet#63] [General]
In order to make QNEthernet Library work on the Teensy 4.1 with the Arduino PubSubClient I had to change the following on line 481 of PubSubClient.cpp

```
//unsigned int rc = 0;
int rc = 0;
```

Once I did that it was pretty straight forward. Here is some example code. I didn't find any other examples of this online so I figured I would capture it here.

```
#include <QNEthernet.h> //for teensy 4.1 https://github.com/ssilverman/QNEthernet?tab=readme-ov-file#ethernetclient
//In order to make this work I had to edit C:\Users\HulkHogan\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp line 481 and change the following
//unsigned int rc = 0; #old line
//int rc = 0;
using namespace qindesign::network;
#include <PubSubClient.h>

IPAddress localIP (192,168,0,3); //static IP
IPAddress subnetMask(255,255,255,0);
IPAddress gateway(192,168,0,1);
IPAddress serverIP (192, 168, 0, 2); //MQTT Broker
const int  MQTT_PORT = 1883;
const char *MQTT_CLIENT_ID = "Hulk";

EthernetClient eth_Client;
PubSubClient MQTT_Client(eth_Client);

void setup() {
  Serial.begin(115200);
  while (!Serial) ; // wait for Arduino Serial Monitor  
  Ethernet.begin(localIP,subnetMask,gateway);
  Serial.println("\nEthernet connected with IP address: ");
  Serial.println(Ethernet.localIP());    
  MQTT_Client.setServer(serverIP,MQTT_PORT);  
  while (!MQTT_Client.connected()) { // Loop until we're reconnected
    if (MQTT_Client.connect(MQTT_CLIENT_ID)) { // Attempt to connect
      Serial.print("MQTT connected");    
    }
    else {
      Serial.println("MQTT connection failed, rc=" + String(MQTT_Client.state()));      
      delay(5000);  // Wait 5 seconds before retrying
    }
  }
}
void loop() {
  MQTT_Client.publish("Catchphrase","Whatcha gonna do, brother, when Hulkamania runs wild on you?",true);
  delay(5000);
}
```

**Comment by ssilverman** [comment]
What was happening before you made that change?

**Comment by kyleti** [comment]
**I got the following error:**

c:\Users\HulkHogan\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp: In member function 'boolean PubSubClient::publish_P(const char*, const uint8_t*, unsigned int, boolean)':
c:\Users\HulkHogan\Documents\Arduino\libraries\PubSubClient\src\PubSubClient.cpp:525:20: warning: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Wsign-compare]
  525 |         return (rc == expectedLength);
      |                 ~~~^~~~~~~~~~~~~~~~~
Opening Teensy Loader...
Memory Usage on Teensy 4.1:
  FLASH: code:140936, data:20428, headers:8616   free for files:7956484
   RAM1: variables:24448, code:137620, padding:26220   free for local variables:336000
   RAM2: variables:55552  free for malloc/new:468736

**Comment by ssilverman** [comment]
I would change `expectedLength` to `unsigned int` instead of changing `rc` to `int`. The warning is probably because of the more recent compiler starting in Teensyduino 1.58.
